src/client.rs

```diff
@@ -280,3 +270,368 @@ where
+    block_hash: T::Hash,
+    ext_hash: T::Hash,
+    client: &'client Client<T>,
+}
+
+impl<'client, T: Config> std::fmt::Debug for TransactionInBlock<'client, T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("TransactionInBlock")
+            .field("block_hash", &self.block_hash)
+            .field("ext_hash", &self.ext_hash)
+            .field("client", &"<client>")
+            .finish()
+    }
+}
+
+impl<'client, T: Config> TransactionInBlock<'client, T> {
+    /// Return the hash of the block that the transaction has made it into.
+    pub fn block_hash(&self) -> T::Hash {
+        self.block_hash
+    }
+
+    /// Return the hash of the extrinsic that was submitted.
+    pub fn extrinsic_hash(&self) -> T::Hash {
+        self.ext_hash
+    }
+
+    /// Fetch the events associated with this transaction. If the transaction
+    /// was successful (ie no `ExtrinsicFailed`) events were found, then we return
+    /// the events associated with it. If the transaction was not successful, or
+    /// something else went wrong, we return an error.
+    ///
+    /// **Note:** If multiple `ExtrinsicFailed` errors are returned (for instance
+    /// because a pallet chooses to emit one as an event, which is considered
+    /// abnormal behaviour), it is not specified which of the errors is returned here.
+    /// You can use [`TransactionInBlock::find_all_events`] if you'd like to handle
+    /// scenarios like this.
+    pub async fn wait_for_success(&self) -> Result<TransactionEvents<T>, Error> {
+        let events = self.find_all_events().await?;
+
+        // Try to find any errors; return the first one we encounter.
+        for ev in events.as_slice() {
+            if &ev.pallet == "System" && &ev.variant == "ExtrinsicFailed" {
+                use codec::Decode;
+                let dispatch_error = sp_runtime::DispatchError::decode(&mut &*ev.data)?;
+                let runtime_error = crate::RuntimeError::from_dispatch(
+                    &self.client.metadata,
+                    dispatch_error,
+                )?;
+                return Err(runtime_error.into())
+            }
+        }
+
+        Ok(events)
+    }
+
+    /// Fetch all of the events associated with this transaction. This succeeds whether
+    /// the transaction was a success or not; it's up to you to handle the error and
+    /// success events however you prefer.
+    pub async fn find_all_events(&self) -> Result<TransactionEvents<T>, Error> {
+        let block = self
+            .client
+            .rpc()
+            .block(Some(self.block_hash))
+            .await?
+            .ok_or(Error::Transaction(TransactionError::BlockHashNotFound))?;
+
+        let extrinsic_idx = block.block.extrinsics
+            .iter()
+            .position(|ext| {
+                let hash = T::Hashing::hash_of(ext);
+                hash == self.ext_hash
+            })
```

Comment:
Are the extrinsics returned in any particular order? Can this be slow (even a DoS vector?) if the block has many extrinsics?

Reply:
Extrinsics are returned in the order that they are added to the block, so we can rely on this position to tell us what the index of the extrinsic was (so that we can work out which events apply to it).

This is quite hideous and I'd like a better API for this personally, but it's the agreed-upon approach for doing this it seems :)

I'd assume that downloading block details can be slower than many calls, and so a possible attack vector for sure.